blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/EnumValidator.java

```diff
@@ -0,0 +1,146 @@
+import com.rackspacecloud.blueflood.types.BluefloodEnumRollup;
+import com.rackspacecloud.blueflood.types.IMetric;
+import com.rackspacecloud.blueflood.types.Locator;
+import com.rackspacecloud.blueflood.types.PreaggregatedMetric;
+import com.rackspacecloud.blueflood.utils.ModuleLoader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.util.*;
+
+/** This class handles collecting enum values for specific metrics, via their locators, and checking whether
+    a uniqueness count of the enum values have reached a certain threshold.  If it has, the class mark the metric
+    as bad by inserting its locator into the proper cassandra column family.  If it hasn't reached the threshold, then it
+    will create or update the elasticsearch "enums" index for the metric.
+**/
+public class EnumValidator implements Runnable {
+
+    private static final Logger log = LoggerFactory.getLogger(EnumValidator.class);
+    private static final Configuration config = Configuration.getInstance();
+    private static final DiscoveryIO discoveryIO = (DiscoveryIO) ModuleLoader.getInstance(DiscoveryIO.class, CoreConfig.ENUMS_DISCOVERY_MODULES);
+    private static final int ENUM_UNIQUE_VALUES_THRESHOLD = config.getIntegerProperty(CoreConfig.ENUM_UNIQUE_VALUES_THRESHOLD);
+    private Set<Locator> locators;
+
+    public EnumValidator(Set<Locator> locators) {
+        this.locators = locators;
+    }
+
+    @Override
+    public void run() {
+
+        if (locators == null) return;
+        Map<Locator, ColumnList<Long>> cfMetrics = AstyanaxReader.getInstance().getStringAndHashMappingForEnums(new ArrayList(locators));
+
+        for (final Locator locator : cfMetrics.keySet()) {
+            // for each locator from CF results, get enum values
+            try {
+                // values from cassandra db
+                Map<Long, String> enumStringValuesFromCF = null;
+                ColumnList<Long> enumHashValuesFromCF = cfMetrics.get(locator);
+                if ((enumHashValuesFromCF != null) && (enumHashValuesFromCF.size() > 0)) {
+                    enumStringValuesFromCF = AstyanaxReader.getInstance().getEnumValueFromHashes(enumHashValuesFromCF);
+                }
+                else {
+                    log.debug(String.format("No enum values found for metric %s", locator.toString()));
+                    // metric has no enum values, skip processing this locator and proceed to next one
+                    continue;
+                }
+
+                // convert enum values to array list
+                ArrayList<String> currentEnumValues = null;
+                if ((enumStringValuesFromCF != null) && (enumStringValuesFromCF.size() > 0)) {
+                    currentEnumValues = new ArrayList<String>(enumStringValuesFromCF.values());
+                }
+
+                // validate enum values count and write to index or bad metric
+                validateThresholdAndWrite(locator, currentEnumValues);
+
```

Comment:
Would it be more efficient to create a list of locators that need to be written and write them all to ES in one batch? Not sure what the performance implications are of doing a separate write for each metric. Do you know?